ffmpeg/ffmpeg.go

```diff
@@ -210,8 +212,23 @@ func newAVOpts(opts map[string]string) *C.AVDictionary {
 	return dict
 }
 
+// check if device number is valid or not.
+func checkDevNum(devNum string) bool {
+	if len(devNum) > 0 {
+		devId, err := strconv.Atoi(devNum)
+		if err != nil || devId > MaxHwDeviceNum {
+			return false
+		}
+	}
+	return true
+}
+
 // return encoding specific options for the given accel
 func configAccel(inAcc, outAcc Acceleration, inDev, outDev string) (string, string, error) {
+	if !checkDevNum(inDev) || !checkDevNum(outDev) {
```

Comment:
I have tried to find a way to catch that ffmpeg error. When I set the invalid device number, the error was occurred from [here]( https://github.com/livepeer/FFmpeg/blob/682c4189d8364867bcc49f9749e04b27dc37cded/libavutil/hwcontext_cuda.c#L399), and it will return `AVERROR_UNKNOWN`. According to the CUDA API, if the device ordinal does not correspond to a valid CUDA device, I might get `CUDA_ERROR_INVALID_DEVICE = 101` error. So we need to modify the code from ffmpeg to get the correct error code. But I think modifying ffmpeg is not good. This should be carefully considered. That's why I am confused on this issue.

Reply:
Were the Nvidia tests failing without this change?

If the Nvidia tests were not failing without this change then I suggest we address this issue separately from this PR.

If the Nvidia tests are failing without this change then I'd suggest exploring the following - in the LPMS C code create a helper that wraps an ffmpeg function for creating a CUDA device. From glancing at the link provided by @JamesWanglf it seems that there is a function for creating a CUDA device and if that returns an error we can return a device creation failure error from the LPMS C code. I'm not sure if there are complications from doing this route though.

Reply:
Yes, actually, it will be failing without this change.
For now, if we meet an unknown error([AVERROR_UNKNOWN](https://github.com/livepeer/lpms/blob/0999e7e2f16c2065a6f6d2eb5b94411bfe911656/ffmpeg/decoder.c#L265)) while running [`av_hwdevice_ctx_create`](https://github.com/livepeer/lpms/blob/0999e7e2f16c2065a6f6d2eb5b94411bfe911656/ffmpeg/decoder.c#L250), it will dig into `panic` statement. But I think this is retryable error, so we don't need to call `panic` for this case.
I think the solution provided by @yondonfu is good.

1. I can define a new error code in LPMS C code(like [this](https://github.com/livepeer/lpms/blob/0999e7e2f16c2065a6f6d2eb5b94411bfe911656/ffmpeg/transcoder.c#L21)), so that we can return this error code if the unknown error occurred when [`av_hwdevice_ctx_create`](https://github.com/livepeer/lpms/blob/0999e7e2f16c2065a6f6d2eb5b94411bfe911656/ffmpeg/decoder.c#L250) function is running.
2. I can also define a new error code in LPMS Go code(in [here](https://github.com/livepeer/lpms/blob/0999e7e2f16c2065a6f6d2eb5b94411bfe911656/ffmpeg/ffmpeg_errors.go#L13)), so that this error will be returned if we meet unknown error, instead of digging into [`panic`](https://github.com/livepeer/lpms/blob/0999e7e2f16c2065a6f6d2eb5b94411bfe911656/ffmpeg/ffmpeg.go#L524).

Reply:
Why not add proper return code to our FFmpeg fork? We might create a PR in the upstream. I think it's a flaw in FFmpeg that it carefully [logs CUDA errors](https://github.com/livepeer/FFmpeg/blob/682c4189d8364867bcc49f9749e04b27dc37cded/libavutil/cuda_check.h#L31), but doesn't propagate what happened into internal error code. We could check return code [here](https://github.com/livepeer/FFmpeg/blob/682c4189d8364867bcc49f9749e04b27dc37cded/libavutil/hwcontext_cuda.c#L399) and, if it's unknown device, re-use some existing FFmpeg return code, like [ENXIO: No such device or address](https://github.com/livepeer/FFmpeg/blob/682c4189d8364867bcc49f9749e04b27dc37cded/libavutil/error.c#L94). At first glace, this will not require any changes to lpms or go-livepeer, aside of tests.